libraries/AP_Compass/AP_Compass_Calibration.cpp

```diff
@@ -0,0 +1,243 @@
+/// -*- tab-width: 4; Mode: C++; c-basic-offset: 4; indent-tabs-mode: nil -*-
+#include <AP_HAL/AP_HAL.h>
+#include "Compass.h"
+#include <AP_Notify/AP_Notify.h>
+
+extern AP_HAL::HAL& hal;
+
+void
+Compass::compass_cal_update()
+{
+    AP_Notify::flags.compass_cal_running = 0;
+
+    for (uint8_t i=0; i<COMPASS_MAX_INSTANCES; i++) {
+        bool failure;
+        _calibrator[i].update(failure);
+        if (failure) {
+            AP_Notify::events.compass_cal_failed = 1;
+        }
+
+        if (_calibrator[i].check_for_timeout()) {
+            AP_Notify::events.compass_cal_failed = 1;
+            cancel_calibration_all();
+        }
+
+        if (_calibrator[i].running()) {
+            AP_Notify::flags.compass_cal_running = 1;
+        }
+    }
+}
+
+bool
+Compass::start_calibration(uint8_t i, bool retry, bool autosave, float delay, bool autoreboot)
+{
+    if (healthy(i)) {
```

Comment:
I would prefer if (!healthy(i)) { return false; } as it makes for easier to read code (less indentation)